File: litecache-python/src/litecache/_serialize.py
"""Value <-> (blob, value_type) conversion.

See SPEC.md for the full type-code table. Codes 0-4 are cross-language
portable: str/int/float/bytes/bool are stored as native UTF-8 text, and
dicts, lists, dataclasses, and plain objects are JSON-encoded. Code 5
(pickle) is a Python-only opt-in escape hatch that is never written unless
the cache is constructed with serializer="pickle". Code 6 is reserved for
other languages' native serializers (e.g. Java); litecache never writes it,
and reading it always raises SerializationError rather than returning raw
bytes silently.

SECURITY: unpickling can execute arbitrary code. litecache never pickles
under serializer="auto" (the default) or serializer="json". Only
serializer="pickle" ever writes pickled data, and reading pickled data back
requires either serializer="pickle" or serializer="auto" with
allow_pickle=True. Treat any cache file that might contain pickled data like
application code: never open one from an untrusted source.
"""

from __future__ import annotations

import dataclasses
import json
import pickle
from typing import Any, get_type_hints

from ._schema import TYPE_BYTES, TYPE_FLOAT, TYPE_INT, TYPE_JAVA, TYPE_JSON, TYPE_PICKLE, TYPE_STR
from .exceptions import SerializationError


def _json_default(value: Any) -> Any:
    if dataclasses.is_dataclass(value) and not isinstance(value, type):
        return dataclasses.asdict(value)
    if hasattr(value, "__dict__"):
        return vars(value)
    raise TypeError(f"object of type {type(value).__name__!r} is not JSON-serializable")


def _encode_json(value: Any) -> bytes:
    text = json.dumps(value, ensure_ascii=False, separators=(",", ":"), default=_json_default)
    return text.encode("utf-8")


def serialize(value: Any, *, mode: str) -> tuple[bytes, int]:
    """Encode a Python value to (blob, value_type) under the given serializer mode.

    mode is one of "auto", "json", "pickle". "auto" and "json" behave
    identically here (both refuse to pickle); "pickle" falls back to
    pickling a value that isn't natively storable or JSON-representable.
    """
    if isinstance(value, bytes):
        return value, TYPE_BYTES
    if isinstance(value, str):
        return value.encode("utf-8"), TYPE_STR
    # bool is a subclass of int; encode it as JSON so it round-trips as a bool.
    if isinstance(value, bool):
        return json.dumps(value).encode("utf-8"), TYPE_JSON
    if isinstance(value, int):
        return str(value).encode("utf-8"), TYPE_INT
    if isinstance(value, float):
        return repr(value).encode("utf-8"), TYPE_FLOAT
    try:
        return _encode_json(value), TYPE_JSON
    except TypeError as exc:
        if mode == "pickle":
            return pickle.dumps(value, protocol=pickle.HIGHEST_PROTOCOL), TYPE_PICKLE
        raise SerializationError(
            f"cannot serialize value of type {type(value).__name__!r}: {exc}. "
            "litecache stores str/int/float/bytes/bool natively and JSON-encodes "
            "dicts, lists, dataclasses, and plain objects (via vars()) otherwise. "
            "Either make this value JSON-representable, or construct "
            "LiteCache(serializer='pickle') to opt into pickling as a last resort."
        ) from exc


def _resolve_type_hints(cls: type) -> dict[str, Any]:
    try:
        return get_type_hints(cls)
    except Exception:
        return {}


def _build_from_cls(data: Any, cls: type) -> Any:
    if not isinstance(data, dict):
        raise SerializationError(
            f"cannot reconstruct {cls.__name__!r} from a stored {type(data).__name__}; "
            "cls reconstruction requires the stored value to be a JSON object"
        )
    kwargs = dict(data)
    if dataclasses.is_dataclass(cls):
        hints = _resolve_type_hints(cls)
        for field in dataclasses.fields(cls):
            if field.name not in kwargs or not isinstance(kwargs[field.name], dict):
                continue
            field_type = hints.get(field.name, field.type)
            if isinstance(field_type, type) and dataclasses.is_dataclass(field_type):
                kwargs[field.name] = _build_from_cls(kwargs[field.name], field_type)
    try:
        return cls(**kwargs)
    except TypeError as exc:
        raise SerializationError(
            f"cannot reconstruct {cls.__name__!r} from stored data {data!r}: {exc}"
        ) from exc


def deserialize(
    blob: bytes,
    value_type: int,
    *,
    mode: str,
    allow_pickle: bool,
    cls: type[Any] | None = None,
) -> Any:
    """Decode a stored (blob, value_type) back to a Python value.

    mode/allow_pickle gate whether a pickled (value_type=5) row may be read:
    "json" mode never allows it; "auto" mode allows it only with
    allow_pickle=True; "pickle" mode always allows it.
    """
    if value_type == TYPE_BYTES:
        return blob
    if value_type == TYPE_STR:
        return blob.decode("utf-8")
    if value_type == TYPE_INT:
        return int(blob.decode("utf-8"))
    if value_type == TYPE_FLOAT:
        return float(blob.decode("utf-8"))
    if value_type == TYPE_JSON:
        data = json.loads(blob.decode("utf-8"))
        if cls is None:
            return data
        return _build_from_cls(data, cls)
    if value_type == TYPE_PICKLE:
        if mode == "json":
            raise SerializationError(
                "this value was stored with pickle (value_type=5), but this LiteCache "
                "was opened with serializer='json', which never reads pickled data; "
                "reopen with serializer='auto' (and allow_pickle=True) or serializer='pickle'"
            )
        if mode == "auto" and not allow_pickle:
            raise SerializationError(
                "this value was stored with pickle (value_type=5); reading it under "
                "serializer='auto' requires allow_pickle=True, or reopen with "
                "serializer='pickle'"
            )
        return pickle.loads(blob)
    if value_type == TYPE_JAVA:
        raise SerializationError(
            "value_type=6 is a Java-native serialization format and cannot be read from Python"
        )
    raise SerializationError(f"unknown value_type {value_type!r} in stored row")

File: lytecache-node/src/serialize.ts
/**
 * Value <-> (bytes, value_type) conversion. See SPEC.md for the full type-code table.
 *
 * Codes 0-4 are cross-language portable: Buffer/string/int/float are stored as native UTF-8
 * text (see the note on INT64/FLOAT64 below), and everything else JSON-encodes. Code 5
 * (Python pickle) and code 6 (Java native serialization) are other languages' escape hatches;
 * this library never writes them, and reading either raises {@link SerializationError} instead
 * of returning raw bytes silently.
 */

import { SerializationError } from "./errors.js";
import { TypeCode, type TypeCodeValue } from "./schema.js";

export interface SerializedValue {
  bytes: Buffer;
  typeCode: TypeCodeValue;
}

export interface DeserializeOptions {
  /** Passed straight through to `JSON.parse` for type-code-4 (JSON) values. */
  reviver?: (key: string, value: unknown) => unknown;
  /**
   * Rehydrates a type-code-4 (JSON) object as an instance of the given class:
   * `Object.assign(Object.create(into.prototype), data)`. The stored data must be a plain
   * object (not an array or primitive).
   */
  into?: new (...args: never[]) => object;
}

const MAX_SAFE = BigInt(Number.MAX_SAFE_INTEGER);
const MIN_SAFE = BigInt(Number.MIN_SAFE_INTEGER);
// int64 bounds, so a stored integer always round-trips through Python's/Java's 64-bit integers.
const INT64_MAX = 9223372036854775807n;
const INT64_MIN = -9223372036854775808n;

function jsonReplacer(_key: string, value: unknown): unknown {
  if (value instanceof Map) {
    throw new SerializationError(
      "cannot serialize a Map -- convert it to a plain object first, e.g. Object.fromEntries(map)",
    );
  }
  if (value instanceof Set) {
    throw new SerializationError(
      "cannot serialize a Set -- convert it to an array first, e.g. Array.from(set)",
    );
  }
  return value;
}

function encodeJson(value: unknown): Buffer {
  let text: string;
  try {
    // undefined here would only happen for values JSON.stringify legitimately drops
    // (nested functions/symbols); a bare `undefined` at the top level is rejected earlier.
    const result = JSON.stringify(value, jsonReplacer);
    if (result === undefined) {
      throw new SerializationError(`cannot serialize a value of type ${typeof value} to JSON`);
    }
    text = result;
  } catch (err) {
    if (err instanceof SerializationError) throw err;
    throw new SerializationError(
      `cannot serialize value to JSON: ${err instanceof Error ? err.message : String(err)}`,
      { cause: err },
    );
  }
  return Buffer.from(text, "utf8");
}

/** Encodes a value to `(bytes, value_type)`. Throws {@link SerializationError} if it can't be. */
export function serialize(value: unknown): SerializedValue {
  if (value === undefined) {
    throw new SerializationError("cannot store undefined -- did you mean null?");
  }
  if (typeof value === "function" || typeof value === "symbol") {
    throw new SerializationError(`cannot serialize a value of type ${typeof value}`);
  }
  if (Buffer.isBuffer(value)) {
    return { bytes: value, typeCode: TypeCode.BYTES };
  }
  if (value instanceof Uint8Array) {
    return { bytes: Buffer.from(value), typeCode: TypeCode.BYTES };
  }
  if (typeof value === "string") {
    return { bytes: Buffer.from(value, "utf8"), typeCode: TypeCode.STR };
  }
  if (typeof value === "bigint") {
    if (value > INT64_MAX || value < INT64_MIN) {
      throw new SerializationError(
        `bigint ${value} is outside the signed 64-bit integer range that can be stored`,
      );
    }
    return { bytes: Buffer.from(value.toString(), "utf8"), typeCode: TypeCode.INT };
  }
  if (typeof value === "number") {
    if (Number.isNaN(value) || !Number.isFinite(value)) {
      throw new SerializationError("cannot store NaN or Infinity");
    }
    if (Number.isInteger(value)) {
      if (value > Number.MAX_SAFE_INTEGER || value < Number.MIN_SAFE_INTEGER) {
        throw new SerializationError(
          `${value} is beyond Number.MAX_SAFE_INTEGER and may have already lost precision; ` +
            "pass a BigInt instead to store it exactly",
        );
      }
      return { bytes: Buffer.from(value.toString(), "utf8"), typeCode: TypeCode.INT };
    }
    return { bytes: Buffer.from(value.toString(), "utf8"), typeCode: TypeCode.FLOAT };
  }
  // Everything else -- plain objects, arrays, booleans, null, Date, class instances (via
  // toJSON() or their enumerable own properties, exactly like JSON.stringify always behaves).
  return { bytes: encodeJson(value), typeCode: TypeCode.JSON };
}

function decodeInt(text: string): number | bigint {
  const big = BigInt(text);
  if (big > MAX_SAFE || big < MIN_SAFE) {
    return big;
  }
  return Number(big);
}

// This library never writes NaN/Infinity (rejected at serialize() time), but Python's
// repr(float('nan'))/repr(float('inf')) is "nan"/"inf" and Java's Double.toString() is
// "NaN"/"Infinity" -- accept all of those spellings on read so a value written by either
// language (however it got there) is still readable rather than silently misparsed.
function decodeFloat(text: string): number {
  const lower = text.toLowerCase();
  if (lower === "nan") return NaN;
  if (lower === "inf" || lower === "infinity") return Infinity;
  if (lower === "-inf" || lower === "-infinity") return -Infinity;
  const value = Number(text);
  if (Number.isNaN(value)) {
    throw new SerializationError(`stored float value is not valid: ${text}`);
  }
  return value;
}

/** Decodes `(bytes, value_type)` back to a value. Throws {@link SerializationError} if it can't be. */
export function deserialize(
  bytes: Buffer,
  typeCode: number,
  options?: DeserializeOptions,
): unknown {
  switch (typeCode) {
    case TypeCode.BYTES:
      return Buffer.from(bytes);
    case TypeCode.STR:
      return bytes.toString("utf8");
    case TypeCode.INT: {
      const text = bytes.toString("utf8");
      try {
        return decodeInt(text);
      } catch (err) {
        throw new SerializationError(`stored integer value is not valid: ${text}`, { cause: err });
      }
    }
    case TypeCode.FLOAT: {
      const text = bytes.toString("utf8");
      try {
        return decodeFloat(text);
      } catch (err) {
        if (err instanceof SerializationError) throw err;
        throw new SerializationError(`stored float value is not valid: ${text}`, { cause: err });
      }
    }
    case TypeCode.JSON: {
      const text = bytes.toString("utf8");
      const data = JSON.parse(text, options?.reviver);
      if (options?.into) {
        if (data === null || typeof data !== "object" || Array.isArray(data)) {
          throw new SerializationError(
            `cannot reconstruct ${options.into.name} from a stored ${Array.isArray(data) ? "array" : typeof data}; ` +
              "'into' rehydration requires the stored value to be a JSON object",
          );
        }
        return Object.assign(Object.create(options.into.prototype) as object, data);
      }
      return data;
    }
    case TypeCode.PYTHON_PICKLE:
      throw new SerializationError(
        "value_type=5 is a Python-pickle-only format and cannot be read from Node.js",
      );
    case TypeCode.JAVA_SERIALIZED:
      throw new SerializationError(
        "value_type=6 is a Java-native serialization format and cannot be read from Node.js",
      );
    default:
      throw new SerializationError(`unknown value_type ${typeCode} in stored row`);
  }
}
